lib/WUI/nhttp/req_parser.cpp

```diff
@@ -0,0 +1,177 @@
+            const size_t key_len = strlen(api_key);
+            if (*idx >= key_len || api_key[*idx] != event.payload) {
+                auth_status = false;
+            } else if (++*idx == key_len) {
+                auth_status = true;
+            }
+        } else {
+            auth_status = false;
+        }
+        return ExecutionControl::Continue;
+    case Names::ContentLength:
+        if (!content_length.has_value()) {
+            content_length = 0;
+        }
+        *content_length = 10 * *content_length + (event.payload - '0');
+        return ExecutionControl::Continue;
+    case Names::Version:
+        switch (event.payload) {
+        case '.':
+            version_major = version_minor;
+            version_minor = 0;
+            return ExecutionControl::Continue;
+        case '0' ... '9':
+            version_minor = 10 * version_minor + (event.payload - '0');
+            return ExecutionControl::Continue;
+        }
+    case Names::ConnectionClose:
+        connection = Connection::Close;
+        break;
+    case Names::ConnectionKeepAlive:
+        connection = Connection::KeepAlive;
+        break;
+    case Names::Body:
+        done = true;
+        // Yes, really don't stop. Eath the \n too!
+        return ExecutionControl::Continue;
+    }
+
+    return ExecutionControl::Continue;
+}
+
+Step RequestParser::step(string_view input, bool terminated_by_client, uint8_t *, size_t) {
+    if (done) {
+        return Step { 0, 0, Continue() };
+    }
+
+    api_key = server->get_api_key();
+    const auto [result, consumed] = consume(input);
+    api_key = nullptr;
+
+    if (!done && result == ExecutionControl::NoTransition) {
+        // Malformed request
+        error_code = Status::BadRequest;
+        done = true;
+    }
+
+    if (!done && terminated_by_client) {
+        // Incomplete request
+        error_code = Status::BadRequest;
+        done = true;
+    }
+
+    if (done) {
+        /*
+         * Note: we assume the array of selectors is terminated by one that
+         * handles unhandled/unknown requests ‒ one that returns 404 or so.
+         */
+        for (const Selector *const *selector = server->selectors();; selector++) {
+            if (auto state = (*selector)->accept(*this); state.has_value()) {
+                return Step { consumed, 0, std::move(*state) };
+            }
+        }
+    } else {
+        return Step { consumed, 0, Continue() };
+    }
+}
+
+bool RequestParser::can_keep_alive() const {
+    return (connection == Connection::KeepAlive) || (version_major == 1 && version_minor >= 1 && connection != Connection::Close);
+}
+
+bool RequestParser::uri_filename(char *buffer, size_t buffer_size) const {
+    // FIXME: Parsing and _validation_ of the URL:
+    // * Stop at any `&`
+    // * Make sure there are no `..` or such in there.
+    // * Demangling?
+    if (buffer_size > url_size) {
+        memcpy(buffer, url.begin(), url_size);
```

Comment:
consider using `strlcpy` instead (makes sure the last character is `\0`)

Reply:
yes, there is an additional implementation elsewhere (copied from BSD) just for the unit tests.